feat(syntax): Add Sapling commit message editor syntax highlighting

vegerot opened the pull request:
feat(syntax): Add Sapling commit message editor syntax highlighting


Summary: Since sl is close enough to hg, just changed `HG` to `SL` or `HG`.

Additionally, to support `sl commit --verbose`, add a `hgcommitDiff.vim`
that's a copy of `syntax/diff.vim` but adds the SL prefix to each line

## Questions for reviewer
- I know copy/pasting `diff.vim` is lame.  Would you rather I simply make the `(HG | SL):` part optional and add it to `diff.vim`?  I only have two remarks about that
  - I don't want to add irrelevant matchers to "diff".vim, and am worried this would set a precedent to further clutter it in the future
  - I haven't measured the performance regression this would cause.  I'm worried it would be noticeable on large files that I personally don't notice in my `sl commit --verbose` editors


Example:
<img width="1062" alt="image" src="https://user-images.githubusercontent.com/5297725/206757125-d2171150-6868-4a62-a886-ca99a2ebbc8e.png">

```hgcommit
feat(syntax): Add Sapling commit message editor syntax highlighting


Summary: Since sl is close enough to hg, just changed `HG` to `SL` or `HG`.

Additionally, to support `sl commit --verbose`, add a `hgcommitDiff.vim`
that's a copy of `syntax/diff.vim` but adds the SL prefix to each line


Test Plan: TODO

 SL: Enter commit message.  Lines beginning with 'SL:' are removed.
 SL: Leave message empty to abort commit.
 SL: --
 SL: user: Max 👨🏽‍💻 Coplan <mchcopl@gmail.com>
 SL: diff --git a/runtime/syntax/hgcommit.vim b/runtime/syntax/hgcommit.vim
 SL: --- a/runtime/syntax/hgcommit.vim
 SL: +++ b/runtime/syntax/hgcommit.vim
 SL: @@ -1,8 +1,9 @@
 SL:  " Vim syntax file
 SL: -" Language:	hg (Mercurial) commit file
 SL: +" Language:	sl/hg (Sapling / Mercurial) commit file
 SL:  " Maintainer:	Ken Takata <kentkt at csc dot jp>
 SL: -" Last Change:	2012 Aug 23
 SL: -" Filenames:	hg-editor-*.txt
 SL: +"  		Max Coplan <mchcopl@gmail.com>
 SL: +" Last Change:	2022-12-08
 SL: +" Filenames:	hgcommit*.vim
 SL:  " License:	VIM License
 SL:  " URL:		https://github.com/k-takata/hg-vim
 SL:
 SL: @@ -10,12 +11,15 @@
 SL:    finish
 SL:  endif
 SL:
 SL: -syn match hgcommitComment "^HG:.*$"             contains=@NoSpell
```

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/vim/vim/pull/11688).
* __->__ #11688

k-takata commented on runtime/syntax/hgcommit.vim:
```diff
@@ -1,21 +1,25 @@
 " Vim syntax file
-" Language:	hg (Mercurial) commit file
+" Language:	sl/hg (Sapling / Mercurial) commit file
```
I would prefer:
```suggestion
" Language:	hg/sl (Mercurial / Sapling) commit file
```

k-takata commented on runtime/syntax/hgcommit.vim:
```diff
@@ -1,21 +1,25 @@
-" Filenames:	hg-editor-*.txt
+"  		Max Coplan <mchcopl@gmail.com>
+" Last Change:	2022-12-08
+" Filenames:	hgcommit*.vim
```
I meant that this syntax file will be applied to `hg-editor-*.txt`. Not sure if `Filenames:` is still needed when this supports sl.

vegerot replied:
```suggestion
```

k-takata commented on runtime/syntax/hgcommit.vim:
```diff
@@ -1,21 +1,25 @@
-syn match hgcommitAdded   "^HG: \zsadded .*$"   contains=@NoSpell contained containedin=hgcommitComment
-syn match hgcommitChanged "^HG: \zschanged .*$" contains=@NoSpell contained containedin=hgcommitComment
-syn match hgcommitRemoved "^HG: \zsremoved .*$" contains=@NoSpell contained containedin=hgcommitComment
+syn match hgcommitComment "^\%(SL\|HG\): .*$"             contains=@NoSpell
```
Fix alignment.
```suggestion
syn match hgcommitComment "^\%(SL\|HG\): .*$"            contains=@NoSpell
```

k-takata commented on runtime/syntax/hgcommit.vim:
```diff
@@ -1,21 +1,25 @@
+syn match hgcommitRemoved "^\%(SL\|HG\): \zsremoved .*$" contains=@NoSpell contained containedin=hgcommitComment
+
+syn region gitcommitDiff start=/\%(^\(SL\|HG\): diff --\%(git\|cc\|combined\) \)\@=/ end=/^\%(diff --\|$\|@@\@!\|[^[:alnum:]\ +-]\S\@!\)\@=/ fold contains=@gitcommitDiff
+syn include @gitcommitDiff syntax/hgcommitDiff.vim
```
Why *git*commitDiff?

vegerot replied:
Because I copy/pasted it :D

k-takata commented on runtime/syntax/hgcommitDiff.vim:
```diff
@@ -0,0 +1,391 @@
+endif
+scriptencoding utf-8
+
+syn match diffOnly	"^\%(SL\|HG\): Only in .*"
```
Isn't it better to use the different group name(s)?
```suggestion
syn match hgdiffOnly	"^\%(SL\|HG\): Only in .*"
```

vegerot replied:
yes.  Will fix

vegerot replied:
done

k-takata commented on runtime/syntax/hgcommitDiff.vim:
```diff
@@ -0,0 +1,391 @@
+" Vim syntax file
+" Language:	Sapling / Mecurial Diff (context or unified)
```
Do we need to support context diff in this file?

vegerot replied:
ngl I don't even know what that means :)

k-takata commented on runtime/syntax/hgcommitDiff.vim:
```diff
@@ -0,0 +1,391 @@
+hi def link diffSubname		PreProc
+hi def link diffComment		Comment
+
+let b:current_syntax = "diff"
```
```suggestion
let b:current_syntax = "hgcommitdiff"
```

vegerot commented on runtime/syntax/hgcommit.vim:
```diff
@@ -1,21 +1,25 @@
+syn match hgcommitRemoved "^\%(SL\|HG\): \zsremoved .*$" contains=@NoSpell contained containedin=hgcommitComment
+
+syn region gitcommitDiff start=/\%(^\(SL\|HG\): diff --\%(git\|cc\|combined\) \)\@=/ end=/^\%(diff --\|$\|@@\@!\|[^[:alnum:]\ +-]\S\@!\)\@=/ fold contains=@gitcommitDiff
+syn include @gitcommitDiff syntax/hgcommitDiff.vim
```
```suggestion
syn region hgCommitDiff start=/\%(^\(SL\|HG\): diff --\%(git\|cc\|combined\) \)\@=/ end=/^\%(diff --\|$\|@@\@!\|[^[:alnum:]\ +-]\S\@!\)\@=/ fold contains=@gitcommitDiff
syn include @hgCommitDiff syntax/hgcommitDiff.vim
```

k-takata commented on runtime/syntax/hgcommit.vim:
```diff
@@ -1,21 +1,24 @@
+syn match hgcommitRemoved "^\%(SL\|HG\): \zsremoved .*$" contains=@NoSpell contained containedin=hgcommitComment
+
+syn region hgCommitDiff start=/\%(^\(SL\|HG\): diff --\%(git\|cc\|combined\) \)\@=/ end=/^\%(diff --\|$\|@@\@!\|[^[:alnum:]\ +-]\S\@!\)\@=/ fold contains=@gitcommitDiff
+syn include @hgCommitDiff syntax/hgcommitDiff.vim
```
```suggestion
syn region hgcommitDiff start=/\%(^\(SL\|HG\): diff --\%(git\|cc\|combined\) \)\@=/ end=/^\%(diff --\|$\|@@\@!\|[^[:alnum:]\ +-]\S\@!\)\@=/ fold contains=@hgcommitDiff
syn include @hgcommitDiff syntax/hgcommitDiff.vim
```

k-takata commented on runtime/syntax/hgcommit.vim:
```diff
@@ -1,21 +1,24 @@
+syn match hgcommitChanged "^\%(SL\|HG\): \zschanged .*$" contains=@NoSpell contained containedin=hgcommitComment
+syn match hgcommitRemoved "^\%(SL\|HG\): \zsremoved .*$" contains=@NoSpell contained containedin=hgcommitComment
+
+syn region hgcommitDiff start=/\%(^\(SL\|HG\): diff --\%(git\|cc\|combined\) \)\@=/ end=/^\%(diff --\|$\|@@\@!\|[^[:alnum:]\ +-]\S\@!\)\@=/ fold contains=@gitcommitDiff
```
You are still using `gitcommitDiff` at the end of this line. Is this intentional?

vegerot replied:
nope :)

vegerot replied:
![image](https://user-images.githubusercontent.com/5297725/207501487-014ff608-f3f4-4282-a628-311119fdc152.png)

vegerot committed:
feat(syntax): Add Sapling commit message editor syntax highlighting


Summary: Since sl is close enough to hg, just changed `HG` to `SL` or `HG`.

Additionally, to support `sl commit --verbose`, add a `hgcommitDiff.vim`
that's a copy of `syntax/diff.vim` but adds the SL prefix to each line

## Questions for reviewer
- I know copy/pasting `diff.vim` is lame.  Would you rather I simply make the `(HG | SL):` part optional and add it to `diff.vim`?  I only have two remarks about that
  - I don't want to add irrelevant matchers to "diff".vim, and am worried this would set a precedent to further clutter it in the future
  - I haven't measured the performance regression this would cause.  I'm worried it would be noticeable on large files that I personally don't notice in my `sl commit --verbose` editors


Example:
<img width="1062" alt="image" src="https://user-images.githubusercontent.com/5297725/206757125-d2171150-6868-4a62-a886-ca99a2ebbc8e.png">

```hgcommit
feat(syntax): Add Sapling commit message editor syntax highlighting


Summary: Since sl is close enough to hg, just changed `HG` to `SL` or `HG`.

Additionally, to support `sl commit --verbose`, add a `hgcommitDiff.vim`
that's a copy of `syntax/diff.vim` but adds the SL prefix to each line


Test Plan: TODO

 SL: Enter commit message.  Lines beginning with 'SL:' are removed.
 SL: Leave message empty to abort commit.
 SL: --
 SL: user: Max 👨🏽‍💻 Coplan <mchcopl@gmail.com>
 SL: diff --git a/runtime/syntax/hgcommit.vim b/runtime/syntax/hgcommit.vim
 SL: --- a/runtime/syntax/hgcommit.vim
 SL: +++ b/runtime/syntax/hgcommit.vim
 SL: @@ -1,8 +1,9 @@
 SL:  " Vim syntax file
 SL: -" Language:	hg (Mercurial) commit file
 SL: +" Language:	sl/hg (Sapling / Mercurial) commit file
 SL:  " Maintainer:	Ken Takata <kentkt at csc dot jp>
 SL: -" Last Change:	2012 Aug 23
 SL: -" Filenames:	hg-editor-*.txt
 SL: +"  		Max Coplan <mchcopl@gmail.com>
 SL: +" Last Change:	2022-12-08
 SL: +" Filenames:	hgcommit*.vim
 SL:  " License:	VIM License
 SL:  " URL:		https://github.com/k-takata/hg-vim
 SL:
 SL: @@ -10,12 +11,15 @@
 SL:    finish
 SL:  endif
 SL:
 SL: -syn match hgcommitComment "^HG:.*$"             contains=@NoSpell
```
